### PR TITLE
ROAD-536 Add VersionJump model and admin integration

### PR DESCRIPTION
**Pull request by arielj**
### Jira Ticket 
https://ombulabs.atlassian.net/browse/ROAD-536

### Motivation / Context

This issue https://github.com/fastruby/points/issues/296
    
### QA / Testing Instructions

- Create a Version Jump object in the admin panel
- Then it can be asigned to projects (new or existing)
- Then, for each version jump, the admin panel will show all the projects associated

### Screenshots:

image
image
image

___

I will abide by the [code of conduct](https://github.com/fastruby/points/blob/main/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: app/models/project.rb [hunk 1]
```diff
@@ -52,6 +53,10 @@ def archived?
     status == "archived"
   end
 
+  def locked?
+    locked_at.present?
+  end
```
Comment (arielj) [inline app/models/project.rb:58]: I can revert this change, but was really simple to add (I updated some code that was using `project.locekd_at.nil?` directly)

File: app/policies/story_policy.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+class StoryPolicy < ApplicationPolicy
+end
```
Comment (arielj) [inline app/policies/story_policy.rb:2]: I had to create all the policies to be able to use this Pundit feature in the admin panel

File: app/views/madmin/application/index.html.erb [hunk 3]
```diff
@@ -11,6 +11,12 @@
         </svg>
       <% end %>
     </form>
+
+    <% if policy(resource).create? %>
+      <%= link_to resource.new_path, class: "bg-white hover:bg-gray-100 text-gray-800 font-semibold py-2 px-4 border border-gray-400 rounded shadow" do %>
+        New <span class="hidden md:inline"><%= resource.friendly_name %></span>
+      <% end %>
+    <% end %>
```
Comment (arielj) [inline app/views/madmin/application/index.html.erb:19]: I took this back from the original gem

Comment (torresga) [reply app/views/madmin/application/index.html.erb:19]: What's the original gem? madmin?

**Comment by torresga** [comment]
@arielj One thing I noticed when testing the admin panel is that there is no way to prevent admins from creating the same version jump multiple times. For example, an admin can create multiple Rails projects that have an initial version of 6.0 and a target version of 6.1. There's also nothing stopping admins from creating a version jump that has more than one jump (think a version jump from 6.0 to 7.1 for example) This might lead to users being confused and possibly disjointed data. I don't know whether we should explore this further, or for our purposes, we can trust our users to not make mistakes. What do you think?

**Comment by arielj** [comment]
> @arielj One thing I noticed when testing the admin panel is that there is no way to prevent admins from creating the same version jump multiple times. For example, an admin can create multiple Rails projects that have an initial version of 6.0 and a target version of 6.1. There's also nothing stopping admins from creating a version jump that has more than one jump (think a version jump from 6.0 to 7.1 for example) This might lead to users being confused and possibly disjointed data. I don't know whether we should explore this further, or for our purposes, we can trust our users to not make mistakes. What do you think?

@torresga I'll add a validation to prevent adding the same jump twice

About a jump of more than 1 version I think it's fine? at least I did that intentionally: I was thinking of using, for example, `Rails / 4.2 - 6.0` for the parent project, and then `Rails / 4.2 - 5.0` `Rails / 5.0 - 5.1`, `Rails / 5.1 - 5.2` and `Rails / 5.2 - 6.0` for each of the sub-projects.

I guess we'll need ernesto to define this better.

File: app/madmin/resources/version_jump_resource.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  # Associations
+  attribute :projects, form: false
+
+  # Uncomment this to customize the display name of records in the admin area.
+  def self.display_name(record)
+    record.to_label
+  end
+
+  # Uncomment this to customize the default sort column and direction.
```
[the remark this answers is not in the document]
Comment (arielj) [reply app/madmin/resources/version_jump_resource.rb:16]: This was already there commented out in the other resources. Not sure why that was not removed before, so I'd say to note remove it just for this resource (maybe some low-priority story can be about cleaning up these comments, I don't want to change files that I don't need to change for a comment)

File: app/views/madmin/application/index.html.erb (hunk 3 above)
Comment (arielj) [reply app/views/madmin/application/index.html.erb:19]: yes, sorry, from this file of madmin https://github.com/excid3/madmin/blob/14599a85f0fb687d6ad9bf30ee38db9f8b1abe3c/app/views/madmin/application/index.html.erb#L15

**Comment by hmdros** [comment]
I would leave a small suggestion to add a badge, maybe a small one inside the box, indicating the [subproject](https://github.com/fastruby/points/pull/299/files#diff-31a7f588fa0b08d2f2cc6ea6c34a6b25125b961b0e9c63bfc77536482878d157R23) version jump:

Captura de tela 2023-09-04 094608
Captura de tela 2023-09-04 093544

**Comment by arielj** [comment]
> I would leave a small suggestion to add a badge, maybe a small one inside the box, indicating the [subproject]

I did that initially, but then I realized this is how we name subprojects and it was redundant to have the same information in the name of the subproject and the label 
image

**Review by torresga — APPROVED** [review]

**Comment by etagwerker** [comment]
> I guess we'll need ernesto to define this better.

@torresga @arielj We can trust that admins will know enough to not create jumps that span multiple versions, so I don't think this is a blocker.

**Comment by arielj** [comment]
Pushed 2 updates:
- version jump column/field for projects in admin panel
image
image

And CSS issue fix:
image

**Comment by etagwerker** [comment]
@arielj Looks good, thank you!

**Pull request merged**